Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp

```diff
@@ -0,0 +1,238 @@
+        auto newStyle = RenderStyle::clone(*style);
+        auto renderer = WebCore::createRenderer<RenderBlockFlow>(RenderObject::Type::BlockFlow, document, WTFMove(newStyle));
+        renderer->initializeStyle();
+        return renderer;
+    };
+
+    auto viewTransitionGroup = createRendererIfNeeded(name, PseudoId::ViewTransitionGroup);
+    auto viewTransitionImagePair = viewTransitionGroup ? createRendererIfNeeded(name, PseudoId::ViewTransitionImagePair) : nullptr;
+    auto viewTransitionOld = viewTransitionImagePair ? createRendererIfNeeded(name, PseudoId::ViewTransitionOld) : nullptr;
+    auto viewTransitionNew = viewTransitionImagePair ? createRendererIfNeeded(name, PseudoId::ViewTransitionNew) : nullptr;
+
+    if (viewTransitionOld)
+        m_updater.m_builder.attach(*viewTransitionImagePair, WTFMove(viewTransitionOld));
+
+    if (viewTransitionNew)
+        m_updater.m_builder.attach(*viewTransitionImagePair, WTFMove(viewTransitionNew));
+
+    if (viewTransitionImagePair)
+        m_updater.m_builder.attach(*viewTransitionGroup, WTFMove(viewTransitionImagePair));
+
+    if (viewTransitionGroup)
+        m_updater.m_builder.attach(*documentElementRenderer.view().viewTransitionRoot(), WTFMove(viewTransitionGroup), beforeChild);
+}
+
+void RenderTreeUpdater::ViewTransition::updatePseudoElementGroup(const RenderStyle& groupStyle, RenderElement& group, RenderElement& documentElementRenderer)
+{
+    auto& documentElementStyle = documentElementRenderer.style();
+    auto name = groupStyle.styleTypeIdentifier();
+
+    auto newGroupStyle = RenderStyle::clone(groupStyle);
+    group.setStyle(WTFMove(newGroupStyle));
+
+    auto createRendererIfNeeded = [&](PseudoId pseudoId) -> RenderPtr<RenderBlockFlow> {
+        auto style = documentElementRenderer.getUncachedPseudoStyle({ pseudoId, name }, &documentElementStyle);
+        if (!style || style->display() == DisplayType::None)
+            return nullptr;
+        auto newStyle = RenderStyle::clone(*style);
+        auto renderer = WebCore::createRenderer<RenderBlockFlow>(RenderObject::Type::BlockFlow, documentElementRenderer.document(), WTFMove(newStyle));
+        renderer->initializeStyle();
+        return renderer;
+    };
+
+    enum class ShouldDeleteRenderer : bool { No, Yes };
+    auto updateRenderer = [&](RenderObject& renderer) -> ShouldDeleteRenderer {
+        auto style = documentElementRenderer.getUncachedPseudoStyle({ renderer.style().styleType(), name }, &documentElementStyle);
+        if (!style || style->display() == DisplayType::None)
+            return ShouldDeleteRenderer::Yes;
+
+        auto newStyle = RenderStyle::clone(*style);
+        downcast<RenderElement>(renderer).setStyle(WTFMove(newStyle));
+        return ShouldDeleteRenderer::No;
+    };
+
+    // Create / remove ::view-transtion-image-pair itself.
+    SingleThreadWeakPtr<RenderElement> imagePair;
+    if ((imagePair = downcast<RenderElement>(group.firstChild()))) {
```

Comment:
This looks weird. Should it be a dynamicDowncast ?

Reply:
Not really, this is null checking `group.firstChild()`. If `group.firstChild()` is present, it's guaranteed to be a `RenderElement`. `uncheckedDowncast` could work here.

Reply:
Maybe we should write the simpler `if (group.firstChild())` and do the downcast in the block ?